docs/features/nullable-reference-types.md

```diff
@@ -22,54 +22,43 @@ namespace System.Runtime.CompilerServices
         AllowMultiple = false)]
     public sealed class NullableAttribute : Attribute
     {
-        public NullableAttribute() { }
-        public NullableAttribute(bool[] b) { }
+        public NullableAttribute(byte b) { }
+        public NullableAttribute(byte[] b) { }
     }
 }
 ```
-The parameter-less constructor is emitted for simple type references with top-level nullability and for type parameter definitions that have a `class?` constraint;
-the constructor with `bool[]` parameter is emitted for type references with nested types and nullability.
+
+Each type reference is accompanied by a NullableAttribute with an array of bytes, where 0 is Oblivious, 1 is NotAnnotated and 2 is Annotated. 
+
+To optimize trivial cases the attribute can be omitted, or instead can be replaced with an attribute that takes a single byte value rather than an array.  
+
+Trivial/optimized cases:
+1)	All parts are NotAnnotated � a NullableAttribute with a single value 1 (rather than an array of 1s)
+2)	All parts are Annotated - a NullableAttribute with a single value 2 (rather than an array of 2s)
+3)	All parts are Oblivious � the attribute is omitted, this matches how we interpret the lack of an attribute in legacy assemblies.
+    For completeness, we would also recognize a NullableAttribute with a single value 0 (rather than an array of 0s),
+    but compiler will never emit an attribute like this. 
+
+NullableAttribute(1) should be placed on a type parameter definition that has a `class!` constraint.
+NullableAttribute(2) should be placed on a type parameter definition that has a `class?` constraint.
+Other forms of NullableAttribute ar not emitted on type parameter definitions and are not specially recognized on them.
+
+The `NullableAttribute` type declaration is synthesized by the compiler if it is not included in the compilation, but is needed to produce the output.
+
 ```c#
 // C# representation of metadata
```

Comment:
It would be good to expand this section with an example involving value types (`int` and `Nullable<int>`).